README.md
# The Tenth Edition UNIX Spell Program

This is the `spell` program from the Tenth Edition Research UNIX
system.  It has been updated somewhat to compile on modern systems,
particularly on Ubuntu Linux.

## Installation

The `makefile` and the code have been modified so that installation
locations are easily changed; editing the `PREFIX` setting in the
`makefile` and then running `make` should be all that's necessary.

The program installs itself under the name `v10spell` to avoid
conflicts with any native spelling checkers already installed or
that might be installed in the future.

After editing the `makefile` (if necessary), it should be enough
to do:

	make
	sudo make install

## History

The original C version of the UNIX `spell` program was written by
Doug McIlroy for V7 UNIX.  The tenth edition version, also by Doug
McIlroy, is apparently a descendant of the original one, but able
to take advantage of the larger address space available on VAXen
as compared to PDP-11s.

Doug McIlroy published a paper on the original UNIX `spell`:

> _Development of a Spelling List_, M.D. McIlroy,
> IEEE Transactions on Communications, volume 30, pp 91-99, 1982.

It's available from his [home page](https://www.cs.dartmouth.edu/~doug/spell.pdf).
For completeness, I have included a copy in the repository.

In private mail to me, Doug says:

> By the 10th edition, the word list no longer needed to be compressed,
> and "part-of-speech" affixability codes had been added, thanks to
> work of Ruby Jane Elliott.  The codes cut down on the acceptance of
> silly derivatives.

However, this version still uses the compressed dictionaries. He
likely has a newer version of the code.

## Changes Made

The original code looks to have been written such that it would
compile for both C and C++.  I have removed the C++ bits and gone
for straight C.  Along the way, I fixed a core dump issue in the
main program that apparently didn't happen on the Research VAX
systems.

In addition, I've made use of modern headers and library bits that
are standard, such as `strdup()` and `EXIT_FAILURE` and `EXIT_SUCCESS`.
In what I think are all the appropriate cases I've replaced `0`
with either `NULL` or `'\0'`.

The original code was very sparing on white space; this seems to
have been Doug's personal style.  I have added more space to make
the code a little bit easier on the eyes.  (At least, on my eyes.)

The shell script driver no longer uses `deroff` or `delatex`.  Neither
of these is standard on Linux systems, and I am not aware of a
version of `deroff` that understands the extended facilities of GNU
`troff`.  Anyone who wants to help repair these deficiencies should
be in touch.

The original code is tagged with `v10-code-from-TUHS`; check out
that tag if you want to see it.

The file `spell-wrapper.sh` was contributed by Nelson H.F. Beebe.
It wraps some additional functionally around the original program.
Read the script for the details.

## Oddities

The `amspell` and `brspell` files differ slightly from those found in
the tenth edition sources tarball.  Nonetheless, they work ok.  (Or at
least, the `amspell` one does.  I have not tested British spelling.)

I did some debugging.  (Debugging code can be found in the `debugging`
branch.) It looks like the word lists used to create the tenth edition
files differed from what's in the tarball.

#### Last Updated

Fri Dec  6 07:51:14 IST 2019


ChangeLog
2019-07-11         Dan Cross             <cross@gajendra.net>

	* Avoid UB by using unsigned types for hashes

2019-12-06         Arnold D. Robbins     <arnold@skeeve.com>

	* code.h (code2str): New function.
	* pcode.c, sprog.c: Replace long with int32_t in strategic
	places to match the code's assumption that long is four bytes.
	This broke spell on 64-bit machines.
	* spell.1: Add a section explaining that spell doesn't always
	ignore case when determining if a word is misspelled or not.
	* spell.pdf: New file.
	* spell-wrapper.sh: New file.
	* makefile (prefix, XCFLAGS): Suggested changes from Nelson Beebe.
	* README.md: Updated.
	* spell.1: Fix a typo.

2019-11-17         Arnold D. Robbins     <arnold@skeeve.com>

	* README.md: Updated.

2019-10-16         Arnold D. Robbins     <arnold@skeeve.com>

	* american: Remove 'prioritize'
	* list: Move 'repressable' from here ...
	* stop: ... to here.
	* README.md: Updated.

2019-09-26         Arnold D. Robbins     <arnold@skeeve.com>

	* README.md: Add reference info for Doug's paper. Thanks
	to him for the info.

2019-09-24         Dan Cross             <cross@gajendra.net>

	* README.md: Formatting cleanups, change `<tt>` tags to
	markdown backticks.
	* sprog.c: Fix all warnings; change to use `getopt(3)`.
	* pcode.c: Fix all warnings.

2019-09-24         Arnold D. Robbins     <arnold@skeeve.com>

	* makefile: Changes for configurability.
	* spellsh.in: Changes for configurability. Also revise the
	code for lack of deroff and to handle a user spelling list.
	* spell.1: Bring in line with Linux -man macros and current code.
	* README.md: New file.

2019-09-22         Arnold D. Robbins     <arnold@skeeve.com>

	* sprog.c: Some more cleanups, move to define for lib directory.
	* makefile: Make bin and lib directories configurable.
	* spellsh.in: Renamed from spell.sh.

2019-09-20         Arnold D. Robbins     <arnold@skeeve.com>

	* ChangeLog: Add early history.
	* pcode.c (main): Add call to setlocale to "C", some more
	code prettifying.
	* sprog.c: Ditto.  And more code "prettifying".

2019-09-17         Arnold D. Robbins     <arnold@skeeve.com>

	* amspell.v10: Renamed from amspell.
	* brspell.v10: Renamed from brspell.
	* .gitignore: New file.
	* sprog.c (readdict): Fix core dump.
	(main): Add int return type.
	(all): Start on code improvements.

2019-09-16         Arnold D. Robbins     <arnold@skeeve.com>

	* pcode.c (wcmp): Use const void * for the arguments.
	Add int return type.
	(main): Add int return type.
	(all): General code improvements: Use of NULL, '\0',
	spacing around operators, etc.

1993-07-10         M. Douglas McIlroy    <doug@research.bell-labs.com>

	Tenth Edition Files. Dates as found in the v10 tarball are:

	-rw-rw-r-- 1 doug doug 343512 Jul 10  1993 list
	-rw-rw-r-- 1 doug doug 163959 Jul 10  1993 amspell
	-rw-rw-r-- 1 doug doug 164027 Jul 10  1993 brspell
	-rw-rw-r-- 1 doug doug  12360 Jul 10  1993 local
	-rw-rw-r-- 1 doug doug  22930 Jul 10  1993 sprog.c
	-rw-rw-r-- 1 doug doug   5844 Jun 30  1993 british
	-rw-rw-r-- 1 doug doug  18518 Jun 30  1993 stop
	-rw-rw-r-- 1 doug doug   5691 Jun 30  1993 american
	-rw-rw-r-- 1 doug doug    712 Jun 12  1993 mkfile
	-rw-rw-r-- 1 doug doug   4741 Jun 11  1993 pcode.c
	-rw-rw-r-- 1 doug doug   2964 Apr 28  1993 spell.1
	-rwxrwxr-x 1 doug doug    912 Mar 10  1993 spell.sh
	-rw-rw-r-- 1 doug doug    617 Feb 12  1993 makefile
	-rw-rw-r-- 1 doug doug   1063 Nov  6  1992 code.h
	-rw-rw-r-- 1 doug doug   1117 Dec 18  1990 README


code.h
#include <string.h>

/*
 * affix codes
 */

#define	ED		(1<<0)	/* +ed, +ing */
#define	ADJ		(1<<1)	/* (nce)-t_ce, +ize,+al, +ness, -t+cy, +ity, +ly */	
#define	NOUN		(1<<2)	/* +s (+es), +make, +hood, +ship +less  */
#define PROP_COLLECT	(1<<3)	/* +'s,  +an, +ship(for -manship) +less */
#define ACTOR		(1<<4)	/* +er  */
#define	EST		(1<<5)
#define COMP		(EST|ACTOR)	/* +er,+est */
#define	DONT_TOUCH	(1<<6)
#define	ION		(1<<7)	/* +ion, +or */
#define	N_AFFIX		(1<<8) 	/* +ic, +ive, +ize, +like, +al, +ful, +ism, +ist, -t+cy, +c (maniac) */
#define	V_AFFIX		(1<<9)	/* +able, +ive, +ity((bility), +ment */
#define	V_IRREG		(1<<10)	/* +ing +es +s*/
#define	VERB		(V_IRREG|ED) 
#define MAN		(1<<11)	/* +man, +men, +women, +woman */
#define	ADV		(1<<12)	/* +hood, +ness */
#define STOP		(1<<14)	/* stop list */
#define	NOPREF		(1<<13)	/* no prefix */

#define MONO		(1<<15)	/* double final consonant as in fib->fibbing */
#define IN		(1<<16) /* in- im- ir, not un- */
#define _Y		(1<<17)	/* +y */

#define ALL		(~(NOPREF|STOP|DONT_TOUCH|MONO|IN))    /*anything goes (no stop or nopref)*/

#ifdef DEBUG
/* code2str --- convert codes to readable string */

static char *
code2str(int code)
{
	static struct codes {
		const char *str;
		int val;
	} codes[] = {
		{ "ED", ED },
		{ "ADJ", ADJ },
		{ "NOUN", NOUN },
		{ "PROP_COLLECT", PROP_COLLECT },
		{ "ACTOR", ACTOR },
		{ "EST", EST },
		{ "DONT_TOUCH", DONT_TOUCH },
		{ "ION", ION },
		{ "N_AFFIX", N_AFFIX },
		{ "V_AFFIX", V_AFFIX },
		{ "V_IRREG", V_IRREG },
		{ "MAN", MAN },
		{ "ADV", ADV },
		{ "STOP", STOP },
		{ "NOPREF", NOPREF },
		{ "MONO", MONO },
		{ "IN", IN },
		{ "_Y", _Y },
		{ NULL, 0 }
	};
	static char buffer[200];
	char *cp = buffer;

	memset(buffer, '\0', sizeof(buffer));
	for (int i = 0; codes[i].str != NULL; i++) {
		if ((code & codes[i].val) != 0) {
			if (cp != buffer)
				*cp++ = '|';
			strcpy(cp, codes[i].str);
			cp += strlen(cp);
		}
	}
	*cp = '\0';

	return buffer;
}
#endif


spell-wrapper.sh
#! /bin/sh -
### ====================================================================
### Add support to v10spell for additional command-line dictionaries
### supplied as options +dict1 +dict2 ... to match BSD/Solaris spell
### syntax.
###
### Usage:
###	v10spell.sh [ +dict1 +dict2 ...] infile(s)
###	v10spell.sh [ +dict1 +dict2 ...] < infile
###
### If A_SPELL points to a file, that file is the first private
### dictionary; otherwise, if $HOME/lib/spelldict is a file, it the
### first one.
###
### All command-line options that begin with a plus sign, and exist as
### files after the + is stripped, are added to the dictionary list.
### The contents of the complete list of dictionaries are then sorted
### into a temporary dictionary with unique words in traditional
### (POSIX or C locale) order, and supplied via A_SPELL to v10spell.
###
### The temporary dictionary is removed on exit or failure.
###
### Caveat: spaces in command line arguments break this script!
###
### Nelson H.F. Beebe, beebe@math.utah.edu
### [26-Sep-2019]
### ====================================================================

IFS='
 	'

PATH="$HOME/bin:/usr/uumath/bin:/usr/local/bin:/usr/bin:/bin"
export PATH

trap '/bin/rm -f $TMPSPELL' EXIT HUP INT QUIT PIPE TERM

TMPSPELL=${TMPDIR-/tmp}/v10spell.tmp.$$

touch $TMPSPELL

DICTFILES=

## If A_SPELL points to a file, it overrides the user default dictionary
for d in $A_SPELL $HOME/lib/spelldict
do
    test -f $d && DICTFILES="$DICTFILES $d" && break
done

ARGS=

for arg in "$@"
do
    case "$arg" in
    +*) DICTFILES="$DICTFILES `echo $arg | sed -e 's/^[+]//'`" ;;
    *)  ARGS="$ARGS $arg" ;;
    esac
done

## Combine all exception dictionaries into a
### unique list in traditional (POSIX, C) order
LC_ALL=C LANG=C sort -u $DICTFILES > $TMPSPELL

A_SPELL=$TMPSPELL v10spell $ARGS


pcode.c
#include <stdio.h>
#include <ctype.h>
#include <locale.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "code.h"

/* read an annotated spelling list in the form
	word <tab> affixcode [ , affixcode ] ...
   print a reencoded version
	octal <tab> word
 */

typedef uint32_t Bits;
typedef	struct Dict Dict;
struct	Dict
{
	char*	word;
	Bits	encode;
};

Dict	words[200000];
char	space[500000];
Bits	encodes[4094];
long	nspace;
long	nwords;
int	ncodes;

void	readinput(FILE*);
int32_t	typecode(char*);
int	wcmp(const void*, const void*);
void	pdict(void);
void	sput(int);

int
main(int argc, char *argv[])
{
	FILE *f;

	setlocale(LC_ALL, "C");

	nwords = 0;
	nspace = 0;
	ncodes = 0;
	if (argc <= 1)
		readinput(stdin);
	while (argc > 1) {
		f = fopen(argv[1], "r");
		if (f == NULL) {
			fprintf(stderr, "Cannot open %s\n", argv[1]);
			exit(EXIT_FAILURE);
		}
		readinput(f);
		fclose(f);
		argc--;
		argv++;
	}
	fprintf(stderr, "words = %ld; space = %ld; codes = %d\n",
		nwords, nspace, ncodes);
	qsort(words, nwords, sizeof(words[0]), wcmp);
	pdict();
	return EXIT_SUCCESS;
}

int
wcmp(const void *a, const void *b)
{

	return strcmp(((Dict*)a)->word, ((Dict*)b)->word);
}

void
readinput(FILE* f)
{
	long i;
	char *code, *bword;
	char line[200];
	long lineno = 0;

	while (fgets(line, sizeof(line), f) != NULL) {
		line[strlen(line)-1] = '\0';
		lineno++;
		code = line;
		while (isspace(*code))
			code++;
		bword = code;
		while (*code && ! isspace(*code))
			code++;

		i = code - bword;
		memcpy(space + nspace, bword, i);
		words[nwords].word = space + nspace;
		nspace += i;
		space[nspace] = '\0';
		nspace++;

		if (*code) {
			*code++ = '\0';
			while (isspace(*code))
				code++;
		}
		words[nwords].encode = typecode(code);
		nwords++;
		if (nwords >= sizeof(words)/sizeof(words[0])) {
			fprintf(stderr, "words array too small\n");
			exit(EXIT_FAILURE);
		}
		if (nspace >= sizeof(space)/sizeof(space[0])) {
			fprintf(stderr, "space array too small\n");
			exit(EXIT_FAILURE);
		}
	}
}


typedef struct	Class
{
	char*	codename;
	int32_t	bits;
} Class;
Class	codea[]  =
{
	{ "a", ADJ },
	{ "adv", ADV },
	{ NULL, 0 },
};
Class	codec[] =
{
	{ "comp", COMP },
	{ NULL, 0 },
};
Class	coded[] =
{
	{ "d", DONT_TOUCH},
	{ NULL, 0 },
};

Class	codee[] =
{
	{ "ed",	ED },
	{ "er", ACTOR },
	{ NULL, 0 },
};

Class	codei[] =
{
	{ "in", IN },
	{ "ion", ION },
	{ NULL, 0 },
};

Class	codem[] =
{
	{ "man", MAN },
	{ "ms", MONO },
	{ NULL, 0 },
};

Class	coden[] =
{
	{ "n", NOUN },
	{ "na", N_AFFIX },
	{ "nopref", NOPREF },
	{ NULL, 0 },
};

Class	codep[] =
{
	{ "pc", PROP_COLLECT },
	{ NULL, 0 },
};
Class	codes[] =
{
	{ "s", STOP },
	{ NULL, 0 },
};

Class	codev[] =
{
	{ "v", VERB },
	{ "va", V_AFFIX },
	{ "vi", V_IRREG },
	{ NULL, 0 },
};

Class	codey[] =
{
	{ "y", _Y },
	{ NULL, 0 },
};

Class	codez[] =
{
	{ NULL, 0 },
};
Class*	codetab[] =
{
	codea,
	codez,
	codec,
	coded,
	codee,
	codez,
	codez,
	codez,
	codei,
	codez,
	codez,
	codez,
	codem,
	coden,
	codez,
	codep,
	codez,
	codez,
	codes,
	codez,
	codez,
	codev,
	codez,
	codez,
	codey,
	codez,
};

int32_t
typecode(char *str)
{
	Class *p;
	long code;
	int n, i;
	char *s, *sp, *st;

	code = 0;

loop:
	for (s = str; *s != '\0' && *s != ','; s++)
		;
	for (p = codetab[*str - 'a']; (sp = p->codename) != NULL; p++) {
		st = str;
		for (n = s - str; ; st++, sp++) {
			if (*st != *sp)
				goto cont;
			n--;
			if (n == 0)
				break;
		}
		code |= p->bits;
		if (*s == '\0')
			goto out;
		str = s + 1;
		goto loop;
	cont:
		;
	}
	fprintf(stderr, "Unknown affix code \"%s\"\n", str);
	return 0;
out:
	for (i = 0; i < ncodes; i++)
		if (encodes[i] == code)
			return i;
	encodes[i] = code;
	ncodes++;
	return i;
}

void
sput(int s)
{

	putchar((s>>8) & 0xff);
	putchar(s & 0xff);
}

void
lput(int32_t l)
{
	putchar((l>>24) & 0xff);
	putchar((l>>16) & 0xff);
	putchar((l>>8) & 0xff);
	putchar(l & 0xff);
}

/*
 * spit out the encoded dictionary
 * all numbers are encoded big-endian.
 *	struct
 *	{
 *		short	ncodes;
 *		int	encodes[ncodes];
 *		struct
 *		{
 *			short	encode;
 *			char	word[*];
 *		} words[*];
 *	};
 * 0x8000 flag for code word
 * 0x7800 count of number of common bytes with previous word
 * 0x07ff index into codes array for affixes
 */

void
pdict(void)
{
	long i, count;
	Bits encode;
	int j, c;
	char *lastword, *thisword, *word;

	sput(ncodes);
	for (i = 0; i < ncodes; i++)
		lput(encodes[i]);

	count = ncodes * 4 + 2;
	lastword = "";
	for (i = 0; i < nwords; i++) {
		word = words[i].word;
		thisword = word;
		for (j = 0; *thisword == *lastword; j++) {
			if (*thisword == '\0') {
				fprintf(stderr, "identical words: %s\n", word);
				break;
			}
			thisword++;
			lastword++;
		}
		if (j > 15)
			j = 15;
		encode = words[i].encode;
		c = (1 << 15) | (j << 11) | encode;
		sput(c);
		count += 2;
		for (thisword = word + j; (c = *thisword) != '\0'; thisword++) {
			putchar(c);
			count++;
		}
		lastword = word;
	}
	fprintf(stderr, "output bytes = %ld\n", count);
}
